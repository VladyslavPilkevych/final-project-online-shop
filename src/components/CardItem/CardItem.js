import React, { useEffect, useState, memo } from 'react';
import PropTypes from 'prop-types';
// import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import styles from './CardItem.module.scss';
import Button from '../Button/Button';
import imageAddToFavIcon from '../../assets/Images/CardItem/addToFavIcon.png';
import imageCart from '../../assets/Images/CardItem/cart.png';
import imageGreenG from '../../assets/Images/CardItem/greenG.png';
import imageInCart from '../../assets/Images/CardItem/inCart.png';
import imageRedPhone from '../../assets/Images/CardItem/redPhone.png';
import imageRemoveFromFavIcon from '../../assets/Images/CardItem/removeFromFavIcon.png';

function CardItem(props) {
  const {
    name, currentPrice, id, img, quantity, previousPrice, elementClassName, model, itemNo,
  } = props;

  // const dispatch = useDispatch();
  // const [previousPrice, setPreviousPrice] = useState(null);
  const [favourite, setFavourite] = useState(false);
  const [inCart, setInCart] = useState(false);
  // useEffect(() => {
  // setpreviousPrice('$' + String(currentPrice.slice(1) * 1.25).split('.')[0] + '.99');
  // }, []);
  const addToFavourite = () => {
    setFavourite(true);
    console.log('addToFavourite');
  };
  const removeFromFav = () => {
    setFavourite(false);
    console.log('removeFromFav');
  };
  const addToCart = () => {
    setInCart(true);
    console.log('addToCart');
  };
  return (
    <div id={id} className={`${styles.productItem} ${elementClassName}`}>
      {quantity
        ? (
          <p className={styles.available}>
            <img className={styles.iconAvailable} alt="icon available" src={imageGreenG} />
            {' '}
            in stock
          </p>
        )
        : (
          <p className={`${styles.available} ${styles.notAvaliable}`}>
            <img className={styles.iconAvailable} alt="icon not available" src={imageRedPhone} />
            {' '}
            check availability
          </p>
        )}
      {
        favourite
          ? <img onClick={removeFromFav} role="presentation" className={styles.iconFav} alt="icon favourite" src={imageRemoveFromFavIcon} />
          : <img onClick={addToFavourite} role="presentation" className={styles.iconFav} alt="icon favourite" src={imageAddToFavIcon} />
      }
      <div className={styles.imageContainer}>
        <Link className={styles.linksToCardPage} to={`/products/${itemNo}`}>
          <img className={styles.imgProduct} alt="product" src={img} />
        </Link>
      </div>
      <Link className={styles.linksToCardPage} to={`/products/${itemNo}`}>
        <h3 className={styles.productName}>{name}</h3>
      </Link>
      <Link className={styles.linksToCardPage} to={`/products/${itemNo}`}>
        <h2 className={styles.productModel}>{model}</h2>
      </Link>
      {previousPrice !== 0 && <span className={styles.previousPrice}>{previousPrice}</span>}
      <span className={styles.price}>{currentPrice}</span>
      <div className={styles.btnCartContainer}>
        {inCart
          ? (
            <Button style={`${styles.btnCart} ${styles.btnInCart}`}>
              <img alt="icon cart" src={imageInCart} />
              <p>Already in cart</p>
            </Button>
          )
          : (
            <Button handleClick={addToCart} style={styles.btnCart}>
              <img alt="icon cart" src={imageCart} />
              <p>Add To Cart</p>
            </Button>
          )}
      </div>
    </div>
  );
}

CardItem.propTypes = {
  /* eslint-disable react/forbid-prop-types */
  // itemContent: PropTypes.object.isRequired,
  /* eslint-enable react/forbid-prop-types */
  itemNo: PropTypes.string.isRequired,
  name: PropTypes.string,
  img: PropTypes.string,
  currentPrice: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  id: PropTypes.string.isRequired,
  quantity: PropTypes.number,
  previousPrice: PropTypes.number,
  elementClassName: PropTypes.string,
  model: PropTypes.string,
};

CardItem.defaultProps = {
  name: '',
  img: '',
  currentPrice: 0,
  quantity: 0,
  previousPrice: 0,
  elementClassName: '',
  model: '',
};

export default memo(CardItem);