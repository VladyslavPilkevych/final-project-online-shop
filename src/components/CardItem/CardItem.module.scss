@mixin fonts($font-family, $size: 16px, $color, $bold: 700) {
    font-family: $font-family, sans-serif;
    font-weight: $bold;
    font-size: $size;
    color: $color;
}
.productItem {
    text-align: left;
    // width: 234px;
    padding-left: 1.5%;
    padding-right: 1.5%;
    border: 2px solid rgb(220, 220, 220);
    background: white;
    overflow: hidden;
    padding-bottom: 20px;
    position: relative;
    // transition: 0.4s linear;
    // transition: all 1s linear;
    // -o-transition: all 1s linear;
    // -webkit-transition: all 1s linear;
    transition: 1s ease;
    .available,
    .iconFav,
    .btnInCart,
    .btnCart {
        display: none;
        opacity: 0;
    }
    .available {
        @include fonts("Poppins", 10px, #78a962, 400);
        line-height: 21px;
        margin-top: 10px;
        .iconAvailable {
            width: 10px;
            height: 10px;
            margin: 5px;
        }
    }
    .notAvaliable {
        color: #c94d3f;
    }
    .iconFav {
        width: 30px;
        height: 30px;
        position: absolute;
        right: 10px;
        top: 15px;
        cursor: pointer;
        z-index: 3;
    }
    .imgProduct {
        transition: 1s ease-in-out;
        display: block;
        // width: 80%;
        height: 150px;
        margin: 5px auto;
        margin-top: 20px;
    }
    .imgProduct:hover {
        transform: scale(1.1);
    }
    h3 {
        @include fonts("Poppins", 13px, black, 400);
        color: #39404b;
        margin: 0;
        margin-top: 15px;
    }
    .oldPrice {
        @include fonts("Poppins", 14px, #666666, 400);
        line-height: 19.6px;
        margin-top: 10px;
        display: block;
        text-decoration: line-through;
    }
    .price {
        @include fonts("Poppins", 18px, black, 600);
        line-height: 25px;
        display: block;
    }
    .btnCart {
        margin: 10px auto;
        margin-bottom: 0;
        background-color: white;
        border: 2px solid #0156ff;
        box-sizing: border-box;
        border-radius: 50px;
        padding: 8px 20px;
        cursor: pointer;
        img {
            width: 19px;
            margin: 2px 10px 2px 0;
        }
        p {
            @include fonts("Poppins", 14px, #0156ff, 600);
            margin-top: 3px;
        }
    }
    .btnInCart {
        margin: 10px auto;
        margin-bottom: 0;
        background-color: white;
        border: 2px solid #C94D3F;
        box-sizing: border-box;
        border-radius: 50px;
        padding: 8px 20px;
        cursor: default;
        p {
            @include fonts("Poppins", 14px, #C94D3F, 600);
            margin-top: 3px;
        }
    }
}
.productItem:hover {
    // position: absolute;
    // z-index: 10;
    .available,
    .iconFav,
    .btnInCart,
    .btnCart {
        display: flex;
        opacity: 1;
        transition: 1s;

        // -webkit-transition: all .5s;
        // -o-transition: all .5s;
        // transition: all .5s;
    }
}
