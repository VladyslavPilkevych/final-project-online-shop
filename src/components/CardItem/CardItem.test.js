import React, { Fragment } from 'react';
// import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
// import { fireEvent, getByText, queryByTestId, queryByText, render, screen } from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';
import CardItem from './CardItem';
// import Router from 'react-router-dom';
// import { Routes, Route, Navigate } from 'react-router-dom';

const itemContent = {
  name: 'testName',
  img: 'testURL',
  currentPrice: 123,
  id: 'testID123',
  quantity: 2,
  previousPrice: 100,
  elementClassName: 'testElementClassName',
  model: 'testModel',
  itemNo: '123',
};

function Component() {
  const {
    name, currentPrice, id, img, quantity, previousPrice, elementClassName, model, itemNo,
  } = itemContent;
  return (
    <BrowserRouter>
      <CardItem
        name={name}
        currentPrice={currentPrice}
        id={id}
        img={img}
        quantity={quantity}
        previousPrice={previousPrice}
        elementClassName={elementClassName}
        model={model}
        itemNo={itemNo}
      />
    </BrowserRouter>
  );
}

describe('Snapshot CardItem should render', () => {
  test('Snapshot CardItem is rendered', () => {
    const { asFragment } = render(<Component />);
    const firstRender = asFragment();
    expect(firstRender).toMatchSnapshot(asFragment());
  });
});

// describe('item add to cart', () => {
//   test('should add to card', () => {
//     const { getByText } = render(<Component />);
//     const button = getByText('Add To Cart');
//     button.click();
//     expect(getByText('Already in cart')).toBeInTheDocument();
//   });
// });
