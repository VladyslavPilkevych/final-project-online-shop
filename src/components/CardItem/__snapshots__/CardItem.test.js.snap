// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot CardItem should render Snapshot CardItem is rendered 1`] = `Object {}`;
